File: src/main/java/ru/bsearch/BSearch4.java
package ru.bsearch;

import org.junit.Test;

import java.util.Collections;
import java.util.List;

import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;

/**
 * User: dima
 * Date: Nov 7, 2010
 */
public class BSearch4 {
    @Test
    public void binarySearch() {
        assertThat(bsearch(1, Collections.<Integer>emptyList()), equalTo(-1));

        assertThat(bsearch(0, asList(1)), equalTo(-1));
        assertThat(bsearch(1, asList(1)), equalTo(0));
        assertThat(bsearch(2, asList(1)), equalTo(-1));

        assertThat(bsearch(0, asList(1, 3)), equalTo(-1));
        assertThat(bsearch(1, asList(1, 3)), equalTo(0));
        assertThat(bsearch(2, asList(1, 3)), equalTo(-1));
        assertThat(bsearch(3, asList(1, 3)), equalTo(1));
        assertThat(bsearch(4, asList(1, 3)), equalTo(-1));

        assertThat(bsearch(0, asList(1, 3)), equalTo(-1));
        assertThat(bsearch(1, asList(1, 2, 3)), equalTo(0));
        assertThat(bsearch(2, asList(1, 2, 3)), equalTo(1));
        assertThat(bsearch(3, asList(1, 2, 3)), equalTo(2));
        assertThat(bsearch(4, asList(1, 2, 3)), equalTo(-1));
    }

    private static int bsearch(int value, List<Integer> values) {
        return bsearch(value, values, 0, values.size());
    }

    private static int bsearch(int value, List<Integer> values, int from, int to) {
        if (from >= to) return -1;

        int midIndex = (from + to) / 2;
        int midValue = values.get(midIndex);
        if (value < midValue) {
            return bsearch(value, values, from, midIndex);
        } else if (value > midValue) {
            return bsearch(value, values, midIndex + 1, to);
        } else if (value == midValue) {
            return midIndex;
        }

        throw new IllegalStateException();
    }
}

File: kotlin/src/katas/kotlin/binarysearch/BSearch4.kt
package katas.kotlin.binarysearch

import datsok.shouldEqual
import org.junit.Test

class BSearch4 {
    @Test fun `find index of an element in a list`() {
        listOf<Int>().binarySearch_(0) shouldEqual -1

        listOf(1).let {
            it.binarySearch_(0) shouldEqual -1
            it.binarySearch_(1) shouldEqual 0
            it.binarySearch_(2) shouldEqual -1
        }

        listOf(1, 2).let {
            it.binarySearch_(0) shouldEqual -1
            it.binarySearch_(1) shouldEqual 0
            it.binarySearch_(2) shouldEqual 1
            it.binarySearch_(3) shouldEqual -1
        }

        listOf(1, 2, 3).let {
            it.binarySearch_(0) shouldEqual -1
            it.binarySearch_(1) shouldEqual 0
            it.binarySearch_(2) shouldEqual 1
            it.binarySearch_(3) shouldEqual 2
            it.binarySearch_(4) shouldEqual -1
        }

        listOf(1, 2, 3, 4).let {
            it.binarySearch_(0) shouldEqual -1
            it.binarySearch_(1) shouldEqual 0
            it.binarySearch_(2) shouldEqual 1
            it.binarySearch_(3) shouldEqual 2
            it.binarySearch_(4) shouldEqual 3
            it.binarySearch_(5) shouldEqual -1
        }
    }

    private fun <E: Comparable<E>> List<E>.binarySearch_(value: E): Int {
        var from = 0
        var to = size

        while (from < to) {
            val midIndex = (from + to) / 2
            val midValue = this[midIndex]

            if (midValue == value) return midIndex
            else if (midValue < value) from = midIndex + 1
            else to = midIndex
        }
        return -1
    }
}
